File: backend/utils/tokenBucket.lua
local key = KEYS[1]

local capacity = tonumber(ARGV[1])
local refill_rate = tonumber(ARGV[2])
local now = tonumber(ARGV[3])

local tokens = redis.call("HGET", key, "tokens")
local last_refill = redis.call("HGET", key, "lastRefillTime")

-- First request: start with a full bucket.
if not tokens then
    tokens = capacity
    last_refill = now
end

tokens = tonumber(tokens)
last_refill = tonumber(last_refill)

-- Elapsed time in seconds.
local elapsed = (now - last_refill) / 1000

-- Refill the bucket.
local new_tokens = math.min(
    capacity,
    tokens + (elapsed * refill_rate)
)

-- The current request is happening now.
last_refill = now

if new_tokens >= 1 then
    new_tokens = new_tokens - 1

    redis.call(
        "HSET",
        key,
        "tokens",
        new_tokens,
        "lastRefillTime",
        last_refill
    )

    -- Keep inactive rate-limit keys from living forever.
    local ttl_seconds = math.ceil(capacity / refill_rate) + 1
    redis.call("EXPIRE", key, ttl_seconds)

    return {
        1,
        math.floor(new_tokens),
        0
    }
end

-- No token available.
redis.call(
    "HSET",
    key,
    "tokens",
    new_tokens,
    "lastRefillTime",
    last_refill
)

local ttl_seconds = math.ceil(capacity / refill_rate) + 1
redis.call("EXPIRE", key, ttl_seconds)

-- Time until one full token becomes available.
local tokens_needed = 1 - new_tokens
local retry_after_seconds = math.ceil(tokens_needed / refill_rate)

return {
    0,
    0,
    retry_after_seconds
}

File: backend/utils/tokenBucket.js
// const bucket = new TokenBucket(10, 2);
class TokenBucket {
  constructor(capacity, refillRate) {
    this.capacity = capacity;       // maximum tokens
    this.refillRate = refillRate;   // tokens per second

    this.tokens = capacity;
    this.lastRefillTime = Date.now();
  }

  consume() {
    const now = Date.now();

    // How many seconds have passed since the last refill
    const elapsedTime = (now - this.lastRefillTime) / 1000;

    // Calculate tokens to add
    const tokensToAdd = elapsedTime * this.refillRate;

    // Refill, but never exceed capacity
    this.tokens = Math.min(
      this.capacity,
      this.tokens + tokensToAdd
    );

    // Update the refill timestamp
    this.lastRefillTime = now;

    // Do we have a token?
    if (this.tokens >= 1) {
      this.tokens -= 1;

      return {
        allowed: true,
        remaining: Math.floor(this.tokens),
      };
    }

    return {
      allowed: false,
      remaining: 0,
    };
  }
}

module.exports = TokenBucket;